### ISSUE TITLE
Struct parameter passed incorrectly without warning on m1

### ISSUE THREAD

**Issue by niklas-ourmachinery** [ziglang/zig#11702]
### Zig Version

0.10.0-dev.2351+b64a1d5ab

### Steps to Reproduce

Set up the following files on a new macOS machine (m1 processor):

**mintest.h**:

~~~c
struct rect {
  float x, y, w, h;
};
~~~

**mintest.zig**:

~~~zig
const std = @import("std");
const print = std.debug.print;

const c = @cImport({
    @cInclude("mintest.h");
});

export fn test_zig(arg_r: c.rect) callconv(.C) void {
    print("ARGS RECEIVED:({d},{d},{d},{d})\n", .{ arg_r.x, arg_r.y, arg_r.w, arg_r.h });
}
~~~

**mintest.c**:

~~~c
#include "mintest.h"
#include <stdio.h>

extern void test_zig(struct rect r);

void test(struct rect r) {
  printf("ARGS SENT     {%g,%g,%g,%g}\n", r.x, r.y, r.w, r.h);
  test_zig(r);
}

int main(int argc, char **argv) {
  struct rect r = {100, 100, 1920, 1000};
  test(r);
}
~~~

**test.sh**

~~~sh
zig build-lib -I. mintest.zig || exit 1
zig cc mintest.c -L. -lmintest -o mintest || exit 1
./mintest
~~~

And run `./test.sh`.

### Expected Behavior

I would either expect this to print an error that struct value parameters of this type are not yet supported (i.e. #1481), or I would expect the printout:

~~~
ARGS SENT     {100,100,1920,1000}
ARGS RECEIVED:(100,100,1920,1000)
~~~

### Actual Behavior

I get the following output:

~~~
ARGS SENT     {100,100,1920,1000}
ARGS RECEIVED:(100,0,100,0)
~~~

**Comment by andrewrk** [comment]
I apologize for this catastrophic user experience.

The good news is that the new compiler shipping with 0.10.0 has finally properly tackled this problem and we have testing infrastructure in place to cover all the issues of this nature. The main focus of the zig project right now is shipping this new compiler. The 0.10.0 GitHub milestone is a measure of progress.

**Comment by Namek** [comment]
Look what happened to my button because of that issue 🤭 Of course, on Windows it's all fine:

https://user-images.githubusercontent.com/31058/173377866-98d53990-8f84-4f1e-ba5e-26aeb43ed913.mov

Instead of `ImVec2{.x=0, .y=0}` it actually passed `ImVec2{.x=0,.y=459}` to ImGUI, no matter what the `y` was set to.

@andrewrk 
Question 1: would `va_list` args also be influenced by this? e.g. calling printf (or ImGUI's `Text()`) with formatted text.
Question 2: is it already fixed on master? I wish to develop on MacOS without waiting for months.

**Comment by ryansname** [comment]
@Namek @niklas-ourmachinery I had also found this problem after a bunch of searching, also with imgui. I also just realised this morning that zig is great for cross compiling and the M1 Macs can run x86 via rosetta. So if you target x86_64-macos-gnu instead of the default, it links and run just fine through rosetta.

It's not optimal, but you can at least develop and run from your M1 Macs.

TLDR:
```
ryan@Ryans-MBP test % cat test.sh 
zig build-lib -I. mintest.zig -target x86_64-macos-gnu || exit 1
zig cc mintest.c -L. -lmintest -o mintest -target x86_64-macos-gnu || exit 1
./mintest

ryan@Ryans-MBP test % sh test.sh 
ARGS SENT     {100,100,1920,1000}
ARGS RECEIVED:(100,100,1920,1000)

ryan@Ryans-MBP test % zig version
0.9.1
```

**Comment by ryoppippi** [comment]
I tested https://github.com/ziglang/zig/issues/11702#issue-1245667719 on 0.10.0-dev.3665+9cf667a21(self-hosted version), and it works correctly on my M1 mac!!

```
ryoppippi in /private/tmp/z C via ↯ v0.10.0-dev.3665+9cf667a21
❯ zig build-lib -I. mintest.zig

ryoppippi in /private/tmp/z C via ↯ v0.10.0-dev.3665+9cf667a21
❯ zig cc mintest.c -L. -lmintest -o mintest

ryoppippi in /private/tmp/z C via ↯ v0.10.0-dev.3665+9cf667a21

ryoppippi in /private/tmp/z C via ↯ v0.10.0-dev.3665+9cf667a21
❯ ./mintest
ARGS SENT     {100,100,1920,1000}
ARGS RECEIVED:(100,100,1920,1000)

ryoppippi in /private/tmp/z C via ↯ v0.10.0-dev.3665+9cf667a21
❯ uname -m
arm64

ryoppippi in /private/tmp/z C via ↯ v0.10.0-dev.3665+9cf667a21
❯ uname -p
arm

ryoppippi in /private/tmp/z C via ↯ v0.10.0-dev.3665+9cf667a21
❯ uname -s
Darwin
```
